lib/ewasm/env.js

```diff
@@ -0,0 +1,117 @@
+const fs = require('fs')
+const path = require('path')
+const BN = require('bn.js')
+const { VmError, ERROR } = require('../exceptions.js')
+
+const transformerRaw = fs.readFileSync(path.join(__dirname, '/system/transform-i64.wasm'))
+const transformerModule = WebAssembly.Module(transformerRaw)
+
+module.exports = class Env {
+  constructor (data) {
+    this._data = data
+    this._results = {
+      gasUsed: new BN(0)
+    }
+
+    // Wasm-js api doesn't yet support i64 param/return values.
+    // https://github.com/WebAssembly/proposals/issues/7
+    this.transformer = WebAssembly.Instance(transformerModule, {
+      'interface': {
+        'useGas': this._useGas.bind(this)
+      }
+    })
+  }
+
+  get imports () {
+    return {
+      ethereum: this.ethereum
+    }
+  }
+
+  get ethereum () {
+    return {
+      getCallValue: this.getCallValue.bind(this),
+      getCallDataSize: this.getCallDataSize.bind(this),
+      callDataCopy: this.callDataCopy.bind(this),
+      useGas: this.transformer.exports.useGas,
+      finish: this.finish.bind(this),
+      revert: this.revert.bind(this)
+    }
+  }
+
+  getCallValue (offset) {
+    const vBuf = this._data.value
+    this._memory.write(offset, 16, vBuf)
+  }
+
+  getCallDataSize () {
+    return this._data.data.length
+  }
+
+  callDataCopy (resultOffset, dataOffset, length) {
+    const data = this._data.data.slice(dataOffset, dataOffset + length)
+    this._memory.write(resultOffset, length, data)
+  }
+
+  _useGas (high, low) {
+    const amount = fromI64(high, low)
+    this.useGas(amount)
+  }
+
+  useGas (amount) {
+    amount = new BN(amount)
+    const gasUsed = this._results.gasUsed.add(amount)
+    if (this._data.gasLimit.lt(gasUsed)) {
+      this._results.exception = 0
+      this._results.exceptionError = ERROR.OUT_OF_GAS
+      this._results.gasUsed = this._data.gasLimit
+      this._results.return = Buffer.from([])
+      throw new VmError(ERROR.OUT_OF_GAS)
+    }
+
+    this._results.gasUsed = gasUsed
+  }
+
+  finish (offset, length) {
+    let ret = Buffer.from([])
+    if (length) {
+      ret = Buffer.from(this._memory.read(offset, length))
+    }
+
+    // 1 = success
+    this._results.exception = 1
+    this._results.return = ret
+  }
+
+  revert (offset, length) {
+    let ret = Buffer.from([])
+    if (length) {
+      ret = Buffer.from(this._memory.read(offset, length))
+    }
+
+    this._results.exception = 0
+    this._results.exceptionError = ERROR.REVERT
+    this._results.gasUsed = this._data.gasLimit
+    this._results.return = ret
+
+    throw new VmError(ERROR.REVERT)
+  }
+
+  setMemory (memory) {
+    this._memory = memory
+  }
+}
+
+// Converts a 64 bit number to two JS numbers
+function fromI64 (high, low) {
+  if (high < 0) {
+    // convert from a 32-bit two's compliment
+    high = 0x100000000 - high
+  }
```

Comment:
Can you add an assert here that `high` is doesn't have any bits set between 32-21 bits?

(JS can store up to 53 bits)